internal/pkg/fieldpropagator/analyzer.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package fieldpropagator implements identification of field propagators.
+// A field propagator is a function that returns a source field.
+package fieldpropagator
+
+import (
+	"go/types"
+	"reflect"
+
+	"github.com/google/go-flow-levee/internal/pkg/config"
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/analysis/passes/buildssa"
+	"golang.org/x/tools/go/ssa"
+)
+
+// ResultType is a map telling whether an object (really a function) is a field propagator.
+type ResultType = map[types.Object]bool
+
+type isFieldPropagator struct{}
+
+func (i isFieldPropagator) AFact() {}
+
+func (i isFieldPropagator) String() string {
+	return "field propagator identified"
+}
+
+// For testing purposes.
+// Specifically, this makes the analyzer report field propagators it finds,
+// which allows us to use analysistest "want" expectations.
+var reporting bool
+
+var Analyzer = &analysis.Analyzer{
+	Name:       "fieldpropagator",
+	Doc:        "This analyzer identifies field propagators.",
+	Flags:      config.FlagSet,
+	Run:        run,
+	Requires:   []*analysis.Analyzer{buildssa.Analyzer},
+	ResultType: reflect.TypeOf(new(ResultType)).Elem(),
+	FactTypes:  []analysis.Fact{new(isFieldPropagator)},
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	ssaInput := pass.ResultOf[buildssa.Analyzer].(*buildssa.SSA)
+
+	conf, err := config.ReadConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	ssaProg := ssaInput.Pkg.Prog
+	for _, mem := range ssaInput.Pkg.Members {
+		if ssaType, ok := mem.(*ssa.Type); ok && conf.IsSource(ssaType.Type()) {
+			methods := ssaProg.MethodSets.MethodSet(ssaType.Type())
+			for i := 0; i < methods.Len(); i++ {
+				meth := ssaProg.MethodValue(methods.At(i))
+
+				// Function does not return anything
+				if res := meth.Signature.Results(); res == nil || (*res).Len() == 0 {
+					continue
+				}
+
+				var lastBlock *ssa.BasicBlock
+				for _, b := range meth.Blocks {
+					if len(b.Succs) == 0 {
+						lastBlock = b
+					}
+				}
+				if lastBlock == nil {
+					continue
+				}
+
+				lastInstr := lastBlock.Instrs[len(lastBlock.Instrs)-1]
+				ret, isRet := lastInstr.(*ssa.Return)
```

Comment:
I think searching for `Return` instructions would be a better first pass.  There may be many.  Then iterate over those instructions / blocks.

Beyond the scope of this PR: what happens in the event of a panic?

Reply:
Agreed. I was going for the simplest possible case, but I don't see a good reason to wait 'til a future PR to do something more generic.

Concerning panics, I don't think I see a non-pathological case where we would need to worry about them, but I may well be missing something. Some thoughts :
* If a panic systematically occurs before the function returns, then I would consider that a programming error and hope that it would be discovered in code review. In any case, it wouldn't be a problem during execution.
* If only some execution paths lead to a panic (with others leading to a source field being returned), then I think it is ok to be conservative and identify the function as a field propagator.
* If there is a "recover" block, we might need to check it too. However, it would be weird for the recover block to lead to a source field being returned, but not any other path.

Reply:
(Specifying it here for clarity: I'm only looking at the last instruction of each block, because according to the documentation a *Return* cannot occur anywhere else.)

Reply:
> (Specifying it here for clarity: I'm only looking at the last instruction of each block, because according to the documentation a Return cannot occur anywhere else.)

This comment indicates that there should be a similar comment in line 89.